File: camera_calibration.py
import numpy as np
import cv2
import glob


 
# prepare object points, like (0,0,0), (1,0,0), (2,0,0) ....,(6,5,0)
objp = np.zeros((7*11,3), np.float32)
objp[:,:2] = np.mgrid[0:11,0:7].T.reshape(-1,2)

objp *= 2
 
# Arrays to store object points and image points from all the images.
objpoints = [] # 3d point in real world space
imgpoints = [] # 2d points in image plane.
 
cap = cv2.VideoCapture(4,cv2.CAP_V4L2)
if not cap.isOpened():
    exit(-1)

cap.set(cv2.CAP_PROP_FRAME_WIDTH,1280)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT,720)

calib_file = cv2.FileStorage("camera_calib.xml", cv2.FileStorage_WRITE)
calib_file.write("cameraWidth", cap.get(cv2.CAP_PROP_FRAME_WIDTH))
calib_file.write("cameraHeight", cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

# termination criteria
criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.001)

key = 0
ret, img = cap.read()
while(ret and key!=27):
    
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    # Find the chess board corners
    ret, corners = cv2.findChessboardCorners(gray, (11,7), None)

    # If found, add object points, image points (after refining them)
    if ret == True:    
        corners2 = cv2.cornerSubPix(gray,corners, (11,11), (-1,-1), criteria)
        # Draw and display the corners
        cv2.drawChessboardCorners(img, (11,7), corners2, ret)
    
    cv2.imshow('img', img)
    key = cv2.waitKey(10)
    if key == 32:
        if len(corners):
            objpoints.append(objp)
            imgpoints.append(corners2)
    
    ret, img = cap.read()
 
cv2.destroyAllWindows()
cap.release()

ret, camera_matrix, dist_coeffs, rvecs, tvecs = cv2.calibrateCamera(objpoints, imgpoints, gray.shape[::-1], None, None)

calib_file.write("cameraMatrix", camera_matrix)
calib_file.write("distortionCoeffs", dist_coeffs)

calib_file.release()

File: sphere_visualizer/camera_calibration.cpp

#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/calib3d/calib3d.hpp>


void createTargetPoints(cv::Size boardSize, cv::Size2f itemSize, std::vector<cv::Point3f>& points)
{
    for(int i = 0; i<boardSize.height; i++)
    {
        for(int j = 0; j<boardSize.width; j++)
        {
            points.push_back(cv::Point3f(j*itemSize.width, i*itemSize.height, 0));
        }
    }
}

int main(int argc, char** argv)
{
    if (argc < 3)
        return -1;

    cv::VideoCapture cap(std::stoi(argv[1]), cv::CAP_V4L2);
    if(!cap.isOpened())
    {
        return -2;
    }

    cv::Size frameSize(1280, 780);
    cv::Size boardSize(11,7);

    cap.set(cv::CAP_PROP_FRAME_WIDTH, frameSize.width);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, frameSize.height);
    cap.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M','J','P','G'));

    frameSize.width = cap.get(cv::CAP_PROP_FRAME_WIDTH);
    frameSize.height = cap.get(cv::CAP_PROP_FRAME_HEIGHT);

    std::vector<cv::Point3f> objectPoints;
    createTargetPoints(boardSize, cv::Size2f(2,2), objectPoints);


    
    cv::Mat frame;
    cv::Mat gray;
    char key = 0;

    std::vector< std::vector< cv::Point3f > > objectsPoints;
    std::vector< std::vector< cv::Point2f > > imgPoints;
    cv::TermCriteria criteria(cv::TermCriteria::EPS | cv::TermCriteria::MAX_ITER, 5, 0.001);

    while(cap.read(frame) && key != 27)
    {
        std::vector<cv::Point2f> corners;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        bool found = cv::findChessboardCorners(gray, boardSize, corners);
        if(found)
        {
            cv::cornerSubPix(gray, corners, cv::Size(11,11), cv::Size(-1,-1), criteria);
            cv::drawChessboardCorners(frame, boardSize, corners, found);
        }
        cv::imshow("capture", frame);
        key = cv::waitKey(10);
        if(key==32)
        {
            if(found)
            {
                imgPoints.push_back(corners);
                objectsPoints.push_back(objectPoints);
            }
        }
    }

    cv::destroyAllWindows();
    cap.release();

    if(!imgPoints.empty())
    {
        cv::Mat cameraMatrix, distCoeffs, rvecs, tvecs;
        
        cv::calibrateCamera(objectsPoints, imgPoints, frameSize, cameraMatrix, distCoeffs, rvecs, tvecs);

        cv::FileStorage calibFile(argv[2],cv::FileStorage::WRITE);

        calibFile.write("cameraWidth", frameSize.width);
        calibFile.write("cameraHeight", frameSize.height);
        calibFile.write("cameraMatrix", cameraMatrix);
        calibFile.write("distortionCoeffs", distCoeffs);

        calibFile.release();

    }

    

    return 0;
}